test/utils/database/query.py

```diff
@@ -1,23 +1,167 @@
+
+    assert len(parsed) == 3
 
 
 def test_insert(db):
     db.cfg.auto_schema = True
 
     # Test insert one
-    db["documents"].insert([{"this": "is a test"}]).execute()
-    result = list(db["documents"].select("this").execute())[0]
+    db["documents"].insert([{"this": "is a test"}])
+    result = db["documents"].select("this").execute()[0]
     assert result["this"] == "is a test"
 
     # Test insert multiple
-    db["documents"].insert([{"this": "is a test"}, {"this": "is a test"}]).execute()
-    results = list(db["documents"].select("this").execute())
+    db["documents"].insert([{"this": "is a test"}, {"this": "is a test"}])
+    results = db["documents"].select("this").execute()
     assert len(results) == 3
 
 
+def test_outputs(db):
+    db.cfg.auto_schema = True
+
+    ids = db['documents'].insert([{'x': i} for i in range(10)])
+    db['_outputs__a__123456789'].insert(
+        [{'_outputs__a__123456789': i + 2, '_source': id} for i, id in enumerate(ids)]
+    )
+    outputs = db['documents'].outputs('a__123456789').execute()
+    print(outputs)
+
+    for r in outputs:
+        assert r['x'] + 2 == r['_outputs__a__123']
+
+
+def test_subset(db):
+    db.cfg.auto_schema = True
+
+    db['documents'].insert([{'x': i} for i in range(10)])
+
+    ids = db['documents'].ids()
+    results = db['documents'].subset(ids[:5])
+
+    pid = db['documents'].primary_id.execute()
+
+    assert set([r[pid] for r in results]) == set(ids[:5])
+
+    db['_outputs__a__123456789'].insert(
+        [{'_outputs__a__123456789': i + 2, '_source': id} for i, id in enumerate(ids)]
+    )
+
+    results = db['documents'].outputs('a__123456789').subset(ids[:5])
+
+    assert set([r[pid] for r in results]) == set(ids[:5])
+
+    assert 'x' in results[0]
+
+
+def test_ids(db):
+    db.cfg.auto_schema = True
+
+    db['documents'].insert([{'x': i} for i in range(10)])
+
+    results = db['documents'].ids()
+
+    assert len(results) == 10
+
+    assert all(isinstance(x, str) for x in results)
+
+
+def test_select_table(db):
+    db.cfg.auto_schema = True
+
+    db['documents'].insert([{'x': i} for i in range(10)])
+
+    results = db['documents'].execute()
+    assert len(results) == 10
+
+
+def test_select_all_cols(db):
+    db.cfg.auto_schema = True
+
+    db['documents'].insert([{'x': i} for i in range(10)])
+
+    q = db['documents'].select()
+
+    results = q.execute()
+
+    assert len(results) == 10
+
+
+def test_select_one_col(db):
+    db.cfg.auto_schema = True
+
+    db['documents'].insert([{'x': i} for i in range(10)])
+
+    q = db['documents'].select('x')
+
+    results = q.execute()
+
+    assert set(results[0].keys()) == {'x'}
+
+
+def test_filter(db):
+    db.cfg.auto_schema = True
+
+    db['documents'].insert([{'x': i} for i in range(10)])
+
+    t = db['documents']
+
+    q = t.filter(t['x'] == 1)
+
+    results = q.execute()
+
+    assert len(results) == 1
+
+    assert results[0]['x'] == 1
```

Comment:
We also need to test multi conditions and difference condition

difference conditions:
1. isin
2. `>`
3. "!="
4. ...

Reply:
Agreed.